internal/pkg/agent/application/upgrade/details/details.go

```diff
@@ -118,6 +119,22 @@ func (d *Details) RegisterObserver(observer Observer) {
 	d.notifyObserver(observer)
 }
 
+// Equals compares the non-lock fields of two Details structs.
+func (d *Details) Equals(otherD *Details) bool {
+	if d == nil && otherD == nil {
+		return true
+	}
+
+	if d == nil || otherD == nil {
+		return false
+	}
```

Comment:
up to you: comparing addresses first to avoid comparing metadata

Reply:
Sorry, I didn't follow what you're suggesting.

Reply:
I think you're suggesting that if the addresses are equal then we can short circuit and return true.  I'll make that change, thanks.

Reply:
Done in 96b824b74e.